Add tracking of goto-function transformation history

thomasspriggs opened the pull request:
This PR adds tracking of goto-function transformation history. This will enable adding checks relating to transform inter-dependencies and checks that appropriate transforms have been performed before reaching symex.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add new transform history data to documentation

This change to the data structure will be implemented in the following
commits in this PR.

thomasspriggs committed:
Add transform history data structure

thomasspriggs committed:
Add unit test of new data structure

thomasspriggs committed:
Add serialisation of transform history to file

So that the record of the transforms which have been applied can be kept
when goto-binaries are passed through multiple invocations of different
tools.

thomasspriggs committed:
Add deserialisation of transform history

This commit includes an INVARIANT on validity of transformations in
history because the history is potentially read from file and
`static_cast`.

thomasspriggs committed:
Add support for reading previous deprecated goto binary version

This commit is intended to be reverted as part of cbmc major release
version 6. The idea is to temporarily include support for the old
version whilst it is needed for transitional reasons. The long term plan
is to remove it in order to remove the maintenance burden of supporting
the old version.

NlightNFotis commented on src/goto-programs/read_bin_goto_object.cpp:
```diff
@@ -254,20 +258,34 @@ bool read_bin_goto_object(
+           "Please recompile."
+        << messaget::eom;
+    }
+    if(version == GOTO_BINARY_VERSION || is_deprecated_version)
```
Is the condition here correct?

It looks like it's still going to try to read in the file if it's a deprecated version, whereas it looks like we're handling the same situation below in the `else if` branch with an error.

thomasspriggs replied:
That is entirely the intended behaviour. The new argument to `read_bin_goto_object` switches between reading the deprecated format and the current format. In this context deprecated means we currently support this, but we are planning to remove the support. Therefore in the case of the deprecated format we warn immediately before reading and then do the appropriate form of reading. In the case of versions older than the deprecated version, the pre-existing error message is reached instead.

thomasspriggs commented:
@tautschnig I have separated the refactoring from this PR into https://github.com/diffblue/cbmc/pull/7634 and https://github.com/diffblue/cbmc/pull/7635

Commit:
Add simple goto_modelt serialization test

Check that a `goto_modelt` containing only one `goto_functiont` element
can be serialized and deserialized correctly.

Commit:
Add test to check goto_historyt is serialized

Add a test checking that `goto_historyt` can be serialized and
deserialized correctly.

Commit:
Add utility functions to manage transform history

Added 2 functions to directly update the `goto_transform_historyt`
of the `goto_functiont` and `goto_functionst`

peterschrammel reviewed (commented):
Where is the history checked?

Commit:
Remove unused declaration of `string_instrumentation` function

This commit removes a transform function from the public header as it is
exposing the transform to `goto_programt` instead of the
`goto_functiont` or above structures. This has been done as the
transformation history is stored at `goto_functiont` level and so it was
not possible to label the function, via the removed API.

kroening commented:
> Where is the history checked?

A history isn't useful for checking, it's informative only. While I appreciate the goal behind this, it may make sense to reconsider the mechanism.

To see why a history isn't useful for enforcement of any particular set of language features, consider that we want to support many flows, beginning with different front-ends. Every time we check, we'd need to allow all of these flows.

Instead of a transformation history, I would advise to use a set of features that are allowed to be present.

TGWDB commented:
> > Where is the history checked?
> 
> A history isn't useful for checking, it's informative only. While I appreciate the goal behind this, it may make sense to reconsider the mechanism.
> 
> To see why a history isn't useful for enforcement of any particular set of language features, consider that we want to support many flows, beginning with different front-ends. Every time we check, we'd need to allow all of these flows.
> 
> Instead of a transformation history, I would advise to use a set of features that are allowed to be present.

I agree with your claim that this will not check that a goto program is in a specific form, or necessarily compatible. However, this does allow a lot of  sanity checking and improvements. In particular this allows:
* A clear record of the transformations performed and in what order.
* Ensuring that a requisite predecessor transformation has been done. Note that this can be lightly implemented for most transformations and thus avoid some issues and mistakes.
* A proxy for checking by requiring a set of transformations to have been completed (e.g. as described [here](https://diffblue.github.io/cbmc/goto-program-transformations.html)).
* Better support for debugging (both of CBMC and by those debugging goto programs).

Note that you were concerned about supporting "many flows", this is straightforward to have by requiring transformations to have acceptable predecessor requirements. Thus, the order is only as strict as necessary, and can support any flows that do not violate the requirements of predecessor transformations.

We considered writing a checker that could examine a goto program and decide if it met specific criteria. However, the complexity and requirements to do this right proved to be larger than we can reasonably commit to. Further, this does not easily allow checking at various intermediate stages, which a history can do very easily.

thomasspriggs commented:
@peterschrammel 

> Where is the history checked?

In the PRs/tickets which come after this one. We divided up the work in our planning so that the first part was to make the information available to write the checks and the following work was to add the checks in. Would it help if we added in an example check to this initial PR?

thomasspriggs committed:
Refactor API for property labeling of single functions

In order to allow updating the history of the labelled function.

thomasspriggs committed:
Add transform_kind value after transformations

Add `goto_transform_kindt` values to `goto_transform_historyt`
of functions after the corrisponding transformation step. This commit
includes the passes in `src/goto-programs` as applied by the `cbmc`
binary.

thomasspriggs committed:
Refactor API used to do lazy assertion to assumption conversion

So that we can add history tracking to it.

peterschrammel commented:
> Would it help if we added in an example check to this initial PR?

No, you can do that in a follow-up PR.

peterschrammel commented on src/goto-programs/write_goto_binary.cpp:
```diff
@@ -100,6 +100,18 @@ static void write_instructions_binary(
   }
 }
 
+static void write_tranform_history_binary(
```
tranform -> transform

thomasspriggs replied:
This should now be fixed.

peterschrammel commented on src/goto-programs/write_goto_binary.h:
```diff
@@ -12,7 +12,7 @@ Author: CM Wintersteiger
 #define CPROVER_GOTO_PROGRAMS_WRITE_GOTO_BINARY_H
 
-#define GOTO_BINARY_VERSION 5
+#define GOTO_BINARY_VERSION 6
```
Is it time to increase CBMC's major version too?

thomasspriggs replied:
There is a plan to increment CBMC's major version later number this year, along with adding further potentially breaking changes. There is an assortment of other issues and PRs which have been tagged with "version 6" here - https://github.com/diffblue/cbmc/labels/Version%206

I have no problem with bumping CBMC's major version number if appropriate, but it may be worth discussing the best way to shepherd in all the version 6 changes before proceeding. It might be that we should just mark the regular releases as "pre-release" until these are all in, but I think the major release process should be agreed first.

peterschrammel commented on src/goto-programs/goto_check.cpp:
```diff
@@ -74,6 +74,8 @@ void transform_assertions_assumptions(
       enable_assertions,
       enable_built_in_assertions,
       enable_assumptions);
+    add_history_transform(
```
Hm... I don't quite like this duplication here. 🔀

thomasspriggs replied:
Neither do I. I'll see what I can do to make this better.

thomasspriggs replied:
This should now be addressed (via a refactor).

peterschrammel commented on src/goto-programs/mm_io.cpp:
```diff
@@ -103,6 +103,7 @@ void mm_io(
       }
     }
   }
+  add_history_transform(goto_transform_kindt::mm_io, goto_function);
```
🔀

thomasspriggs replied:
The duplication in this transform would have actually added to the history twice. Therefore I have straight removed one of the duplicates. :facepalm:

peterschrammel commented on src/goto-programs/set_properties.cpp:
```diff
@@ -103,6 +103,7 @@ void label_properties(goto_model_functiont &function)
   auto &goto_function = function.get_goto_function();
   label_properties(
     function.get_function_id(), goto_function.body, property_counters);
+  add_history_transform(goto_transform_kindt::label_properties, goto_function);
```
🔀

thomasspriggs replied:
This should now be addressed via a refactor.

thomasspriggs committed:
Deduplicate calls to `add_history_transform` in `goto_check.cpp`

The shared implementation function needs to be passed the
`goto_function` instead of the `goto_program` in order to carry out this
deduplication because the `goto_program` does not include the history.

thomasspriggs committed:
Mark implementation overload of `label_properties` as `static`

Because this helps make it clear that this particular overload is not
exposed in the header.

thomasspriggs committed:
Deduplicate call to `add_history_transform` in `label_properties`

The shared implementation function needs to be passed the
`goto_function` instead of the `goto_program` in order to carry out this
deduplication because the `goto_program` does not include the history.

kroening reviewed (changes requested):
Please do not merge -- I'd like to discuss this in detail.

tautschnig commented:
@TGWDB said:

> * A clear record of the transformations performed and in what order.
> * Ensuring that a requisite predecessor transformation has been done. Note that this can be lightly implemented for most transformations and thus avoid some issues and mistakes.
> * A proxy for checking by requiring a set of transformations to have been completed (e.g. as described [here](https://diffblue.github.io/cbmc/goto-program-transformations.html)).
> * Better support for debugging (both of CBMC and by those debugging goto programs).

I somewhat agree (caveats below) that the above is being accomplished, but is this actually what we need? To a consumer of a goto program, does the/should the order of transformations matter? Does it actually matter whether a certain transformation took place, or isn't it really the resulting goto program that matters? Does it perhaps make matters worse when one knows the descriptions of transforms, but not necessarily what the resulting features are?

Take function pointers as an example of a concept that many analyses cannot deal with. They would, therefore, want to check that the transformation history includes function-pointer removal. This implies that:
1) Such a check also needs to know about all the other transforms that may (re-)introduce function pointers. This makes our code base closed in that no one could build a new tool on top of our code base if said tool included transforms that add function pointers; well, the mere fact that we have that enum over transforms within our code base already makes it impossible to extend it.
2) Such a check needs to know all the types of function pointer removals, of which we have multiple. The current approach addresses this problem by putting all those transforms into one basket, which means we talk about transforms, but actually really mean their result.
3) If a producer of goto programs knows that function pointers cannot occur (perhaps they aren't part of the input language), then they would have to spuriously add `goto_transform_kindt::remove_function_pointers` even though no such transform had actually taken place.

I'd prefer listing the ingredients ("may contain nuts") rather than listing the production process.

jimgrundy commented:
To me the latest comments bring this conversation back to beginning, to where the request that triggered this PR began. We know that some goto constructs are allowed only some of the time. Various transformations add/remove some constructs or require some constructs to be/no-be present (ok, mostly not be), and all this is not well documented and not widely understood. I wanted this corrected because we trip over it.

Ideally we would have some number of defined "forms" of goto programs, an understanding of which constructs are allowed/not-allowed in which forms, and for each transformation knowledge of the input and output form. I'd like there to be executable predicates for the forms. A function that takes a goto program and returns true iff it is in some form. I'd use these in assertions (disabled in production because they would mostly be slow) to confirm we had this right.

The feedback I got was that this was not a practical request today. It's a fair bit of code, but worse, the knowledge to do it doesn't really exist and the amount of archeology/experimentation required to build the knowledge to write the code I wanted was too great ... and so this offer of just recording the transformations that had been performed on a goto program for debugging was made. This is not satisfying, but I was ok with it as a step towards what I wanted ... not a very direct step, but doable and it would build some of the knowledge required to get to where I wanted to go, and produce some inspectable data that could perhaps be mined to produce more of the knowledge.

TGWDB commented:
Thanks for all the thoughtful comments and responses.

> I somewhat agree (caveats below) that the above is being accomplished, but is this actually what we need? 

There seems to be a collection of things that different groups want or need. So the answer to this question depends on the groups involved. Here I wish to reiterate that this does not attempt to resolve all concerns/needs for all groups.

> To a consumer of a goto program, does the/should the order of transformations matter? Does it actually matter whether a certain transformation took place, or isn't it really the resulting goto program that matters?

This again depends on the "consumer". We currently seem to have dependencies of the form of "transform X should come after transform Y". This could also be captured by "transform X assumes that the goto program does [not] contain feature C". However, checking that a goto model does [not] have a feature is significantly more expensive to write/check than looking at the transformation history. As mentioned by Jim, a goto program checker for all of these features would be great, but this is a heavy and complete solution as opposed to a quick sanity checker.

> Does it perhaps make matters worse when one knows the descriptions of transforms, but not necessarily what the resulting features are?

I see this (particularly as implemented here) as adding information. At worst this could let someone better see and understand what's going on?

> Take function pointers as an example of a concept that many analyses cannot deal with. They would, therefore, want to check that the transformation history includes function-pointer removal.

To me this is already jumping from "we can use this new information to prevent obvious mistakes" to "we expect this new information to allow us to validate correctness of usage".

> This implies that:
> 
> 1. Such a check also needs to know about all the other transforms that may (re-)introduce function pointers. This makes our code base closed in that no one could build a new tool on top of our code base if said tool included transforms that add function pointers;

I'd argue this is too strong a request for what is in scope. If a transform requires that function pointers are removed then a possible sanity check is that the "function pointer removal" transform is in the history (and we can also add other checks for other transforms we know about). Then if a new transform is added and things break, the information is in the history and a clue is there that function pointers may be the culprit.

> well, the mere fact that we have that enum over transforms within our code base already makes it impossible to extend it.

Fair point. The current implementation of history is not easily externally extensible.


> 2. Such a check needs to know all the types of function pointer removals, of which we have multiple. The current approach addresses this problem by putting all those transforms into one basket, which means we talk about transforms, but actually really mean their result.

Yes. The argument here is that it is better to sanity check and record, than to do nothing (assuming we don't have the time/resources to implement proper validation).

> 3. If a producer of goto programs knows that function pointers cannot occur (perhaps they aren't part of the input language), then they would have to spuriously add `goto_transform_kindt::remove_function_pointers` even though no such transform had actually taken place.

Yes, if we take your proposed approach of doing a sanity check over the whole history. However, we envisaged the first sanity checks being confirming the order of transforms is that currently used/expected in CBMC. This would mean it is an error to call them out of order. With that in mind, if you add a new producer of goto programs it only needs to either: A) lie about the last transform it did to be compatible; or B) add it's last transform as a valid predecessor to the following transform.

> I'd prefer listing the ingredients ("may contain nuts") rather than listing the production process.

I think most of us would love to have a complete list of what is[n't] valid at each stage, but it's questionable that anyone knows this, let alone had code to validate it.

kroening commented:
> The feedback I got was that this was not a practical request today.

It's not that hard. You take the code of the transforms, and instead of letting them transform, you return `true` whenever they found anything to transform.

tautschnig commented:
[...]
> There seems to be a collection of things that different groups want or need. So the answer to this question depends on the groups involved. Here I wish to reiterate that this does not attempt to resolve all concerns/needs for all groups.
> 

Should we perhaps start with some form of document or statement that describes these (possibly conflicting) desires? Good old Requirements Engineering?

> > To a consumer of a goto program, does the/should the order of transformations matter? Does it actually matter whether a certain transformation took place, or isn't it really the resulting goto program that matters?
> 
> This again depends on the "consumer". We currently seem to have dependencies of the form of "transform X should come after transform Y". This could also be captured by "transform X assumes that the goto program does [not] contain feature C". However, checking that a goto model does [not] have a feature is significantly more expensive to write/check than looking at the transformation history. As mentioned by Jim, a goto program checker for all of these features would be great, but this is a heavy and complete solution as opposed to a quick sanity checker.
> 

Do we really have "transform X should come after transform Y?" I'd argue that's a bug and needs to be fixed locally. What I'd expect: transform Y requires that the input be in a certain shape (with no constraints on how you arrived at that shape). I'll state here: I don't think that a check that a goto model does [not] have a feature is a prerequisite. It might be nice to build that at some point in the future, but it is not a required first step.

[...]
> > Take function pointers as an example of a concept that many analyses cannot deal with. They would, therefore, want to check that the transformation history includes function-pointer removal.
> 
> To me this is already jumping from "we can use this new information to prevent obvious mistakes" to "we expect this new information to allow us to validate correctness of usage".
> 

I'm not sure why? (But also this is a minor point in what is becoming a long conversation, so feel free to ignore.)

> > This implies that:
> > 
> > 1. Such a check also needs to know about all the other transforms that may (re-)introduce function pointers. This makes our code base closed in that no one could build a new tool on top of our code base if said tool included transforms that add function pointers;
> 
> I'd argue this is too strong a request for what is in scope. If a transform requires that function pointers are removed then a possible sanity check is that the "function pointer removal" transform is in the history (and we can also add other checks for other transforms we know about). Then if a new transform is added and things break, the information is in the history and a clue is there that function pointers may be the culprit.
> 

It seems I should first clarify what I thought was the most important requirement: have a means to tell whether a given goto program is usable as input to an analysis or transformation, because, as Jim put it, "we trip over it" (and https://github.com/diffblue/cbmc/pull/7605 is a recent such example). This is _not_ addressed by 'If a transform requires that function pointers are removed then a possible sanity check is that the "function pointer removal" transform is in the history,' because a transform that happened after function pointer removal may have re-introduced function pointers. (And it's exacerbated by the fact that not all transforms add themselves to the history, which also makes a bit questionable whether the proposed solution delivers the necessary clues.)

[...]
> > 2. Such a check needs to know all the types of function pointer removals, of which we have multiple. The current approach addresses this problem by putting all those transforms into one basket, which means we talk about transforms, but actually really mean their result.
> 
> Yes. The argument here is that it is better to sanity check and record, than to do nothing (assuming we don't have the time/resources to implement proper validation).
> 

I am not suggesting we do nothing. But I'll try to be more explicit in what I think we should do, below.

> > 3. If a producer of goto programs knows that function pointers cannot occur (perhaps they aren't part of the input language), then they would have to spuriously add `goto_transform_kindt::remove_function_pointers` even though no such transform had actually taken place.
> 
> Yes, if we take your proposed approach of doing a sanity check over the whole history. However, we envisaged the first sanity checks being confirming the order of transforms is that currently used/expected in CBMC. This would mean it is an error to call them out of order. With that in mind, if you add a new producer of goto programs it only needs to either: A) lie about the last transform it did to be compatible; or B) add it's last transform as a valid predecessor to the following transform.
> 

See my earlier comment: I think it's an error if anything assumes a particular order of transforms. Which is not to say that all orderings are permitted, but the order should be derived from the language features permitted to be present in a goto program.

> > I'd prefer listing the ingredients ("may contain nuts") rather than listing the production process.
> 
> I think most of us would love to have a complete list of what is[n't] valid at each stage, but it's questionable that anyone knows this, let alone had code to validate it.

1) I don't think we need a complete list right from the start. I actually believe the current enum of transforms is a good starting point, but we should have an enum of language features, not transforms. (E.g., we should talk about vectors, unions, function pointers instead of "remove_vector", "remove_union", "remove_function_pointers".)
2) I'd expect goto-program producers and transformers to declare what language features may be present in their output goto programs. Automatically inferring these is nice to have, but not blocking this effort in any way.
3) I don't know whether it's better to have a positive set (may_have_vectors, may_have_unions, may_have_function_pointers) or a negative set (has_no_vectors, has_no_unions, has_no_function_pointers). The latter might be easier to get right, because a safe overapproximation is to just clear the set (in the positive case a safe overapproximation requires knowing all the features we consider at this time). We could go as far as making `goto_programt::update` clear that set of flags so that we have a simple way to get started.

So I think the design changes I am suggesting are:
1) Don't keep a history, just record the current state.
2) Turn the recording-of-transforms into recording-of-absence (I think that's what it mostly is, hardly ever is it "presence") of language features.
3) There is no need for perfection: using the current list of transforms as an indication of relevant language features seems like a very good first step for me.

kroening commented:
On the "positive list" vs "negative list". Suppose you wanted to introduce a new language feature to GOTO binaries (e.g., something from Rust or Go or Java). You then have to re-make all binaries to say that they don't use it, even if they come from an entirely unrelated language.

TGWDB commented:
## Goal
The broad goal of various parties involved is to be able to ensure that the GOTO model given to goto-symex is suitable to be symexed. Let’s call such a GOTO model “core-GOTO”.
A side goal is to ensure that calls to GOTO transforms are consistent with the respective GOTO transforms’ pre- and postconditions.

## Approaches
There are two general approaches to achieve this:
A. Check whether a GOTO model is core-GOTO.
B. Check whether transforms have been applied to a GOTO model that imply that it is core-GOTO.

## Constraints

1. No-one currently knows what all the transforms do.
2. No-one currently knows the acceptable input forms for all the transforms.
3. No-one knows all the acceptable orderings that transforms can be called in.
4. There are many GOTO models that are modified by and between transforms and there are no defined/canonical forms.
5. Engineering efforts are limited (time/budget).
6. Future extensions need to be considered.
7. Solution should be able to avoid certain known pitfalls related to GOTO model forms and transforms

## Proposals
### Proposal 1 - Positive checks

Based on approach A.
Core-GOTO and GOTO transforms pre- and postconditions are defined in terms of which features the GOTO models are allowed to include.
The GOTO model's expressions themselves are checked for compliance.

**Pros:**
1.P.1 Clear definition, strongest guarantees (7)
1.P.2 New features are rejected by default and require explicit decisions to allow them (6, 7)
1.P.3 Removing features has no immediate impact (6)
1.P.4 Problematic modifications to transforms may be detected (6, 7)

**Cons:**
1.C.1 Expensive to implement right away (5)
1.C.2 Difficult to engineer incrementally as it is expensive to define relevant over-approximations that can be refined later (need to figure out and enumerate or reasonably abstract features, 1-4, 5)

### Proposal 2 - Negative checks

Based on approach A.
Core-GOTO and GOTO transforms pre- and postconditions are defined in terms of which features the GOTO models are NOT allowed to contain.
The GOTO model's expressions themselves are checked for compliance.

**Pros:**
2.P.1 Quite clear definition, quite strong guarantees (7)
2.P.2 Removing features has no immediate impact (6)
2.P.3 Problematic modifications to transforms may be detected (6, 7)
2.P.4 Can be engineered incrementally as over-approximations can be defined (5)

**Cons:**
2.C.1 New features are allowed by default and require extra care not to forget to implement new checks for them (6, 7)
2.C.2 High effort to figure out which features should be excluded (1-4, 5)
2.C.3 Does not provide a path towards Proposal 1 (6)

### Proposal 3 - Negative status annotations

Based on approach A.
Core-GOTO and GOTO transforms pre- and postconditions are defined in terms of which feature annotations the GOTO models are NOT allowed to contain.
The GOTO model is annotated with abstracted features, which are checked for compliance.

Approach suggested by Michael Tautschnig (with an alternative consideration using positive status annotations).

**Pros:**
3.P.1 Provides rough guarantees wrt pre- and postconditions of transforms (7)
3.P.2 Moderate effort to engineer (5)
3.P.3 Fairly little knowledge about transforms required (1-4)
3.P.4 Provides a path towards Proposal 2 in future (6)

**Cons:**
3.C.1 Relies on transforms to declare the correct annotations in their pre- and postconditions (7)
3.C.2: Annotations are stored in goto binaries, which need to be rebuilt when the annotations change (6)
3.C.3 New features are allowed by default and require extra care (6, 7)

### Proposal 4 - Transform history annotations

Based on approach B. 
GOTO models are annotated with an ordered history list of labels, denoting a sequence of transforms that have been applied to the GOTO model.
The precondition check of a transform checks whether the last transform in the history is a valid predecessor.

This is #7633 .

**Pros:**
4.P.1 Provides strong guarantees wrt the currently known sequence of transforms (7)
4.P.2 Easy to engineer (5)
4.P.3 No detailed knowledge about transforms required (1-4)

**Cons:**
4.C.1 Relies on transforms to record the correct annotations and that a known sequence of transforms is correct (7)
4.C.2 Annotations are stored in goto binaries, which need to be rebuilt when transforms are added or removed (6)
4.C.3 Alternative transforms with the same outcome must be explicitly added to the transform predecessor checks (6)
4.C.4 New features are allowed by default and require extra care (6, 7)

TGWDB commented:
We'd like to converge on a conclusion about this PR (i.e. Proposal 4 above) fairly soon if possible. Clearly it will not solve all the issues and is not the ideal/perfect solution, but under the constraints listed I believe there is a strong argument this adds value in the near future.

kroening commented:
I'm in favour of "positive feature list"!

thomasspriggs commented:
I am closing this out in favour of alternative approaches.
